src/jit/codegenarm64.cpp

```diff
@@ -6536,161 +6432,293 @@ CodeGen::genIntrinsic(GenTreePtr treeNode)
+                unsigned  structOffset  = 0;
+                var_types nextType      = type0;
+
+                // Use the ldp instruction for a struct that is exactly 16-bytes in size
+                //             ldp     x2, x3, [x0]
+                //
+                if (remainingSize == 2*TARGET_POINTER_SIZE)
+                {
+                    if (hasGCpointers)
+                    {
+                        // We have GC pointers, so use two ldr instructions
+                        //
+                        // We do it this  way because we can't currently pass or track 
+                        // two different emitAttr values for a ldp instruction.
+
+                        // Make sure that the first load instruction does not overwrite the addrReg.
+                        //
+                        if (loReg != addrReg)
+                        {
+                            emit->emitIns_R_R_I(INS_ldr, emitTypeSize(type0), loReg, addrReg, structOffset);
+                            emit->emitIns_R_R_I(INS_ldr, emitTypeSize(type1), hiReg, addrReg, structOffset + TARGET_POINTER_SIZE);
+                        }
+                        else 
+                        {
+                            assert(hiReg != addrReg);
+                            emit->emitIns_R_R_I(INS_ldr, emitTypeSize(type1), hiReg, addrReg, structOffset + TARGET_POINTER_SIZE);
+                            emit->emitIns_R_R_I(INS_ldr, emitTypeSize(type0), loReg, addrReg, structOffset);
+                        }
+                    }
+                    else
+                    {
+                        // Use a ldp instruction 
+
+                        emit->emitIns_R_R_R_I(INS_ldp, EA_PTRSIZE, loReg, hiReg, addrReg, structOffset);
+                    }
+                    remainingSize = 0;     // We completely wrote the 16-byte struct
+                }
+
+                regNumber curReg = loReg;
+                while (remainingSize > 0)
+                {
+                    if (remainingSize >= TARGET_POINTER_SIZE)
+                    {
+                        remainingSize -= TARGET_POINTER_SIZE;
+
+                        if ((curReg != addrReg) || (remainingSize == 0))
+                        {
+                            emit->emitIns_R_R_I(INS_ldr, emitTypeSize(nextType), curReg, addrReg, structOffset);
+                        }
+                        else
+                        {
+                            deferLoad = true;
+                            deferAttr = emitTypeSize(nextType);
+                            deferOffset = structOffset;
+                        }
+                        curReg = hiReg;
+                        structOffset += TARGET_POINTER_SIZE;
+                        nextType = type1;
+                    }
+                    else // (remainingSize < TARGET_POINTER_SIZE)
+                    {
+                        int loadSize = remainingSize;
```

Comment:
This code seems to assume that the size must either be 1, 2 or 4. What about a struct that had three fields of long, int and short - wouldn't that leave a remainingSize of 6? Is there some reason that and other combinations don't occur? If so, it would be good to describe that here.